Source/Core/Core/HW/MemmapFunctions.cpp

```diff
@@ -239,94 +220,66 @@ __forceinline void WriteToHardware(u32 em_address, const T data, Memory::XCheckT
 			return;
 		}
 	}
-	else if (((em_address & 0xF0000000) == 0x80000000) ||
-		((em_address & 0xF0000000) == 0xC0000000) ||
-		((em_address & 0xF0000000) == 0x00000000))
+	else if (segment == 0x8 || segment == 0xC || segment == 0x0)
 	{
 		*(T*)&m_pRAM[em_address & RAM_MASK] = bswap(data);
 		return;
 	}
-	else if (m_pEXRAM && (((em_address & 0xF0000000) == 0x90000000) ||
-		((em_address & 0xF0000000) == 0xD0000000) ||
-		((em_address & 0xF0000000) == 0x10000000)))
+	else if (m_pEXRAM && (segment == 0x9 || segment == 0xD || segment == 0x1))
 	{
 		*(T*)&m_pEXRAM[em_address & EXRAM_MASK] = bswap(data);
 		return;
 	}
-	else if ((em_address >= 0xE0000000) && (em_address < (0xE0000000+L1_CACHE_SIZE)))
+	else if (segment == 0xE && (em_address < (0xE0000000+L1_CACHE_SIZE)))
 	{
 		*(T*)&m_pL1Cache[em_address & L1_CACHE_MASK] = bswap(data);
 		return;
 	}
-	else if ((bFakeVMEM && ((em_address &0xF0000000) == 0x70000000)) ||
-		(bFakeVMEM && ((em_address &0xF0000000) == 0x40000000)))
-	{
-		// fake VMEM
-		*(T*)&m_pFakeVMEM[em_address & FAKEVMEM_MASK] = bswap(data);
-	}
 	else
 	{
-		// MMU
-		// Handle stores that cross page boundaries (ewwww)
-		if (sizeof(T) > 1 && (em_address & (HW_PAGE_SIZE-1)) > HW_PAGE_SIZE - sizeof(T))
+translateaddress:
+		if (bFakeVMEM && (segment == 0x7 || segment == 0x4))
 		{
-			T val = bswap(data);
-			u32 tlb_addr = TranslateAddress(em_address, flag);
-			for (u32 addr = em_address; addr < em_address + sizeof(T); addr++, tlb_addr++)
-			{
-				if (!(addr & (HW_PAGE_SIZE-1)))
-					tlb_addr = TranslateAddress(addr, flag);
-				if (tlb_addr == 0)
-				{
-					if (flag == FLAG_WRITE)
-					{
-						if (!SConfig::GetInstance().m_LocalCoreStartupParameter.bMMU)
-							PanicAlertT("Invalid Write to 0x%08x, PC = 0x%08x ", em_address, PC);
-						else
-							GenerateDSIException(addr, true);
-						break;
-					}
-				}
-				else
-				{
-					if (m_pEXRAM && (tlb_addr & 0xF0000000) == 0x10000000)
-					{
-						m_pEXRAM[tlb_addr & EXRAM_MASK] = (u8)val;
-						val >>= 8;
-					}
-					else
-					{
-						m_pRAM[tlb_addr & RAM_MASK] = (u8)val;
-						val >>= 8;
-					}
-				}
-			}
+			// fake VMEM
+			*(T*)&m_pFakeVMEM[em_address & FAKEVMEM_MASK] = bswap(data);
+			return;
 		}
-		else
+
+		// MMU: Do page table translation
+		u32 tlb_addr = TranslateAddress<flag>(em_address);
+		if (tlb_addr == 0)
```

Comment:
Unlikely to affect games we support in bMMU mode; the page 0x0 contains interrupt-handling code, so games wouldn't use it for paging.  It might become an issue when we implement bBAT mode.